sched/scheduler/jobRunner.go

```diff
@@ -0,0 +1,267 @@
+package scheduler
+
+import (
+	"fmt"
+	"github.com/scootdev/scoot/saga"
+	"github.com/scootdev/scoot/sched"
+	cm "github.com/scootdev/scoot/sched/clustermembership"
+	dist "github.com/scootdev/scoot/sched/distributor"
+	"sync"
+)
+
+const MAX_RETRY = 3
+
+type jobRunner struct {
+	job       sched.Job
+	saga      saga.Saga
+	sagaState *saga.SagaState
+	dist      *dist.PoolDistributor
+
+	updateChan chan stateUpdate
+	wg         sync.WaitGroup
+}
+
+func NewJobRunner(job sched.Job, saga saga.Saga, sagaState *saga.SagaState, nodes []cm.Node) jobRunner {
+	jr := jobRunner{
+		job:        job,
+		saga:       saga,
+		sagaState:  sagaState,
+		dist:       dist.NewPoolDistributor(cm.StaticClusterFactory(nodes)),
+		updateChan: make(chan stateUpdate),
+	}
+
+	go jr.updateSagaState()
+	return jr
+}
+
+// Runs the Job associated with this JobRunner to completion
+func (jr jobRunner) runJob() {
+	// don't re-run an already completed saga
+	if jr.sagaState.IsSagaCompleted() {
+		return
+	}
+
+	// if the saga has not been aborted, just run each task in the saga
+	// TODO: this can be made smarter to run the next Best Task instead of just
+	// the next one in order etc....
+	if !jr.sagaState.IsSagaAborted() {
+		for _, task := range jr.job.Tasks {
+
+			if !jr.sagaState.IsTaskCompleted(task.Id) {
+				node := jr.dist.ReserveNode()
+
+				jr.updateChan <- stateUpdate{
+					msgType: saga.StartTask,
+					taskId:  task.Id,
+				}
+
+				jr.wg.Add(1)
+				go func(node cm.Node, task sched.Task) {
+
+					// TODO: After a number of attempts we should stop
+					// Trying to run a task, could be a poison pill
+					// Implement deadletter queue
+					taskExecuted := false
+					for !taskExecuted {
+						err := node.SendMessage(task)
```

Comment:
The current API and localNode implementation block until the message has completed processing.    The idea here is it would just block, once it returns a response (right now success or err) we know its completed and can log EndTask